### PR TITLE
feat: Add button to clear value from search fields

### LINKED ISSUE #21686: Search Bar Clear Button
## **Summary**
These have been designed now: 

Screen Shot 2021-12-19 at 5 47 46 PM

Screen Shot 2021-12-19 at 5 47 46 PM

## **Acceptance Criteria**
When text has been entered into a search input, a custom "clear" X button should appear on the right hand size of the input. If there is a "number of matches" result count, the X appears to the left of it.

This seems like something that should happen on the Inline Specs List as well.

┆Issue is synchronized with this [Jira Task](https://cypress-io.atlassian.net/browse/UNIFY-682) by [Unito](https://www.unito.io)
┆Attachments: <a href="https://cypress-io.atlassian.net/rest/api/2/attachment/content/11022">Screen Shot 2021-12-19 at 5.43.13 PM.png</a> | <a href="https://cypress-io.atlassian.net/rest/api/2/attachment/content/11023">Screen Shot 2021-12-19 at 5.47.46 PM.png</a>
┆author: Mark Noonan
┆epic: UI/UX Improvements
┆friendlyId: UNIFY-682
┆priority: Lowest
┆sprint: Backlog
┆taskType: Task

### PR DESCRIPTION
**Pull request by mike-plummer**
- Closes #21686 

### User facing changelog
* Adds new button to clear search field when filtering specs
* Fix styling issue if a very long search value is input on the Specs List page

### Additional details
Refactoring the `Input` component to rely on flexbox for layout of prefix/suffix, was previously using absolute positioning. This caused issues with the filter field in SpecsList- the search value would underlap the suffix section because it had a dynamic width based on the number of specs

### Steps to test
1. Open project into E2E or CT mode
2. On Specs List view, type a value into the search field. Note that a new "X" button appears on the right side immediately left of the result count section
3. Click the "X" button. Observe field is cleared, spec list resets, and "X" button hides
4. Verify that new "X" button is navigable by keyboard and has appropriate `aria-label`
5. Input a very long value into the field, observe that the value scrolls rather than underlaps result count section
6. Navigate into a spec, open inline spec search
7. Type a value into the search field, observe new "X" button appears on right side
8. Click "X" button. Observe field is cleared, spec list resets, and "X" button hides
9. Verify that new "X" button is navigable by keyboard and has appropriate `aria-label`
10. Verify other usages of the `Input` component still style appropriately. Examples:
  * Project filter field in global mode
  * Filename field in "New Spec" modal
 
### How has the user experience changed?

https://user-images.githubusercontent.com/11482842/172703636-237745a7-48bd-4ddb-ae8f-403c749446d3.mov

### PR Tasks
- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by jennifer-shehane** [comment]
@mike-plummer Have you looked through the percy snapshots? (Do you have access?) Is this flake or has something changed in the css that's affecting the icons/text on many of the other components?

**Comment by mike-plummer** [comment]
> @mike-plummer Have you looked through the percy snapshots? (Do you have access?) Is this flake or has something changed in the css that's affecting the icons/text on many of the other components?

@jennifer-shehane Good catch! About half looks like the normal flake, but my changes have shifted the prefix/suffix section of the input just a pixel or two more than Percy likes. I'll look at getting those shifted back

**Comment by rachelruderman** [comment]
@mike-plummer so sorry for the late review! It's been an eventful sprint on firewatch 🔥 😅 I'll try to get to this today!

**Review by ZachJW34 — APPROVED** [review]
Works and look great!

File: packages/app/src/specs/InlineSpecListHeader.vue [hunk 1]
```diff
@@ -47,6 +47,18 @@
...
+        @click.stop="clearInput"
+      >
+        <i-cy-delete_x16
+          class="icon-light-gray-1000 group-hocus:icon-dark-indigo-300"
+          :class="inputFocused ? 'icon-dark-indigo-300' : 'icon-dark-gray-800'"
+        />
+      </button>
     </div>
     <button
       tabindex="-1"
```
Comment (marktnoonan) [inline packages/app/src/specs/InlineSpecListHeader.vue:64]: Not part of this PR, but while we're in here, let's remove this `tabindex="-1"` that makes this New Spec button unreachable with the keyboard, I have no idea why it's there.

**Review by marktnoonan — APPROVED** [review]
Love the description and steps to test including accessibility!

File: packages/app/src/specs/SpecsListHeader.cy.tsx [hunk 2]
```diff
@@ -4,27 +4,47 @@ import { defaultMessages } from '@cy/i18n'
...
+    setup () {
+      return () => h(SpecsListHeader, {
+        modelValue: search.value,
+        'onUpdate:modelValue': (val: string) => {
+          search.value = val
+        },
+      })
+    },
+  }))
+}
```
Comment (rachelruderman) [inline packages/app/src/specs/SpecsListHeader.cy.tsx:22]: How come you're using a search ref as opposed to just `cy.get`? Was it not working as expected?

Comment (rachelruderman) [reply packages/app/src/specs/SpecsListHeader.cy.tsx:22]: Ah maybe because you need the `modelValue` prop?

Comment (rachelruderman) [reply packages/app/src/specs/SpecsListHeader.cy.tsx:22]: Altho I'd think `modelValue` would just update on its own in accordance with the component behavior... haha sorry stream of consciousness over here

File: packages/app/src/specs/SpecsListHeader.vue [hunk 3]
```diff
@@ -14,7 +14,16 @@
...
-          class="rounded-r-md outline-none h-38px mr-[-0.75rem] group"
+          v-if="props.modelValue"
+          data-cy="clear-search-button"
+          class="border-transparent rounded-md flex outline-none h-24px mr-16px w-24px duration-300 items-center justify-center group hocus-default hocus:ring-0 hocus:bg-indigo-50"
+          :aria-label="t('specPage.clearSearch')"
+          @click="clearInput"
+        >
+          <i-cy-delete_x16 class="icon-light-gray-50 icon-dark-gray-500 group-hocus:icon-dark-indigo-500" />
+        </button>
+        <button
```
Comment (rachelruderman) [inline packages/app/src/specs/SpecsListHeader.vue:25]:
```suggestion
        <button
          type="button"
```

File: packages/app/src/specs/InlineSpecListHeader.cy.tsx [hunk 4]
```diff
@@ -40,6 +39,20 @@ describe('InlineSpecListHeader', () => {
...
   })
 
+  it('clears search field when clear button is clicked', () => {
+    mountWithResultCount(0)
+
+    cy.findByTestId('clear-search-button')
+    .should('not.exist')
+
+    cy.findByLabelText(defaultMessages.specPage.searchPlaceholder)
+    .type('abcd', { delay: 0, force: true })
```
[the remark this answers is not in the document]
Comment (mike-plummer) [reply packages/app/src/specs/InlineSpecListHeader.cy.tsx:49]: In this particular case the label overlays the input, so the force is necessary to overcome that

File: packages/frontend-shared/src/locales/en-US.json [hunk 5]
```diff
@@ -119,6 +119,7 @@
     "pageTitle": "Specs",
     "newSpecButton": "New Spec",
     "searchPlaceholder": "Search Specs",
+    "clearSearch": "Clear search pattern",
```
Comment (rachelruderman) [inline packages/frontend-shared/src/locales/en-US.json:122]: "search field" or "search input" sounds more natural to me than "search pattern". Maybe I've been looking at code too much xD

File: packages/app/src/specs/InlineSpecListHeader.cy.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (rachelruderman) [reply packages/app/src/specs/InlineSpecListHeader.cy.tsx:49]: would you mind adding a short comment to line 49 explaining that?

File: packages/frontend-shared/src/locales/en-US.json (hunk 5 above)
Comment (mike-plummer) [reply packages/frontend-shared/src/locales/en-US.json:122]: You're not wrong, I struggled a bit with the wording here. I agree that "Search Field" makes more sense
